README.md
# inference_server

This package is a wrapper between ROS and the [Tensorflow Object Detection API](https://github.com/tensorflow/models/tree/master/research/object_detection). The Tensorflow Object Detection API is an open source framework built on top of TensorFlow that makes it easy to construct, train and deploy object detection models. This package helps us to use the object detection along with ROS in the form of an action_server. 

## Features:
   * ROS-TensorFlow Object Detection API Wrapper
   * Available as a ROS action_server
   * ROS service to change the inference model
   * Desired class(es) selection with the parameter server

## ROS API
### actions
   * Inference.action
   
   ```
	---
	sensor_msgs/CompressedImage image
	int16 num_detections
	string[] classes
	float32[] scores
	sensor_msgs/RegionOfInterest[] bounding_boxes
	---
   ```

## Testing it with the Robot 
  * Run the object detection action_server with the following command:
 ```
roslaunch inference_server inference_server.launch
 ```
  * To run inference on the image published from the robot, run the action_client and send the goal using the following command:
```
rosrun actionlib axclient.py /inference_server inference_server/InferenceAction inference_server/InferenceAction
```
  * To choose your desired classes for inference, set the parameter server with the following command (By default, it is set to 'all' classes):
```
# For example, if we want to detect keyboard and bicycle alone:
rosparam set /inference_server/desired_classes ['keyboard','bicycle']
```
  * Change the inference model using rosservice as shown:
```
rosservice call /inference_server/change_inference_model "model_name: 'ssd_inception_v2_coco_11_06_2017'
reset_desired_classes_param: true"
```
  * Receive the result with the following fields from the inference of the image, in a chronological order of the detection score.
	* image - Resultant image after inference from Object Detection API
	* num_detections - Number of detected objects in the inference image
	* classes - name of the class to which the object belongs (depends on the model used for the inference)
	* scores - detection scores or the confidence of the detection of the particular object as a particular class
	* bounding_boxes - bounding box of each of the detected object



CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package inference_server
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Forthcoming
-----------
* Merge branch 'optional_image' into 'master'
  added an optional image as an input to Inference action
  See merge request saikishor/inference_server!8
* added an optional image as an input to Inference action
* Merge branch 'fixing_cv_bgr' into 'master'
  Fixing CV Image BGR Issue
  See merge request saikishor/inference_server!7
* removing unused cvbridge and sys headers
  (cherry picked from commit 110a3ee9e6616692199fff438190ff4b609e13aa)
* fixing BGR issue with cv2
  (cherry picked from commit 15e362f962b04bb2ad28cac0b125e968c638d395)
* Merge branch 'change_model_service' into 'master'
  add service to update with new inference model
  See merge request saikishor/inference_server!5
* add service to update with new inference model
* Contributors: Sai Kishor Kothakota, Victor Lopez, saikishor

0.0.1 (2018-08-09)
------------------
* Merge branch 'inference_time' into 'master'
  Inference_time and subscriber changes
  See merge request saikishor/inference_server!4
* reverting subscriber changes
* added inference_time logging
* Merge branch 'desired_classes_param' into 'master'
  Setting desired classes using param server
  See merge request saikishor/inference_server!3
* README.md Update
* added support to choose the desired classes from param server
* Merge branch 'locations_as_params' into 'master'
  Locations as params
  See merge request saikishor/inference_server!2
* fixing the unavailable model download location
* passing inference details through param server
* Merge branch 'add-tensorflow-link' into 'master'
  Update README.md
  See merge request saikishor/inference_server!1
* Update README.md
* Removed the unnecessary subscription of images and added timeout for receiving images
* update README.md
* add README.md
* correcting bounding_boxes typo in action
* removed unused headers
* object_detection class update and inference_server_ndoe update
* added RegionofInterest action msg and initial inference image
* InferenceAdvanced Action and Object Detection Class update
* add gpu allow_growth config
* inference_server initial commit
* reset repository
* image action_server
* Contributors: Victor Lopez, saikishor


src/obj_det_fun.py
import numpy as np
import os
import six.moves.urllib as urllib
import sys
import tarfile
import tensorflow as tf
import zipfile
import rospy
import getpass 

from collections import defaultdict
from io import StringIO
from matplotlib import pyplot as plt
from PIL import Image

#username = getpass.getuser()
TF_MODELS_PATH = '/TFM/models/research/object_detection'
MODEL_DATABASE = '/TFM/model_database'
		
# This is needed to access files from research folder of Tensorflow models API
sys.path.append(TF_MODELS_PATH)	

from utils import label_map_util

from utils import visualization_utils as vis_util
	
# What model to download or use.
MODEL_NAME = 'ssd_inception_v2_coco_11_06_2017'
#MODEL_NAME = 'tiago_model_200000'
MODEL_FILE = MODEL_NAME + '.tar.gz'
DOWNLOAD_BASE = 'http://download.tensorflow.org/models/object_detection/'

#local path directory of all models
MODEL_PATH = os.path.join(MODEL_DATABASE, MODEL_FILE)

# Path to frozen detection graph. This is the actual model that is used for the object detection.
PATH_TO_CKPT = MODEL_DATABASE + '/' + MODEL_NAME + '/frozen_inference_graph.pb'

# List of the strings that is used to add correct label for each box.
PATH_TO_LABELS = os.path.join(TF_MODELS_PATH + '/data', 'mscoco_label_map.pbtxt')
#PATH_TO_LABELS = os.path.join(TF_MODELS_PATH + '/data', 'tiago_label_map.pbtxt')

NUM_CLASSES = 90
#NUM_CLASSES = 1

def loadmodel():
	if os.path.exists(PATH_TO_CKPT):
	    rospy.logwarn('Found and verified .pb file : ' +  PATH_TO_CKPT)	
	    return
	if os.path.exists(MODEL_PATH):
	    rospy.logwarn('Found and verified : ' +  MODEL_PATH)
	    tar_file = tarfile.open(MODEL_PATH)
	else:
	    rospy.logwarn('The model is not available at the model database location: '+ MODEL_PATH)
	    rospy.logwarn('Attempting to download over internet:' + MODEL_FILE) 
	    opener = urllib.request.URLopener()
	    opener.retrieve(DOWNLOAD_BASE + MODEL_FILE, MODEL_FILE)
	    rospy.logwarn('\nDownload Complete!')
	    tar_file = tarfile.open(MODEL_FILE)
	for file in tar_file.getmembers():
	  file_name = os.path.basename(file.name)
	  if 'frozen_inference_graph.pb' in file_name:
	    #choose this option only if you want to extract in current work directory cwd
#		    tar_file.extract(file, os.getcwd())
	    tar_file.extract(file, MODEL_DATABASE)

def objdetect():
	detection_graph = tf.Graph()
	with detection_graph.as_default():
	  od_graph_def = tf.GraphDef()
	  with tf.gfile.GFile(PATH_TO_CKPT, 'rb') as fid:
	    serialized_graph = fid.read()
	    od_graph_def.ParseFromString(serialized_graph)
	    tf.import_graph_def(od_graph_def, name='')
	label_map = label_map_util.load_labelmap(PATH_TO_LABELS)
	categories = label_map_util.convert_label_map_to_categories(label_map, max_num_classes=NUM_CLASSES, use_display_name=True)
	category_index = label_map_util.create_category_index(categories)
	return detection_graph, category_index

def load_image_into_numpy_array(image):
#	(im_width, im_height) = image.size
	(im_width, im_height, _) = image.shape
#	return np.array(image.getdata()).reshape((im_height, im_width, 3)).astype(np.uint8)
	return np.array(image).reshape((im_height, im_width, 3)).astype(np.uint8)

def detect(sess, detection_graph,category_index, data):
	IMAGE_SIZE = (12, 8)
	with detection_graph.as_default():
		# Definite input and output Tensors for detection_graph
		image_tensor = detection_graph.get_tensor_by_name('image_tensor:0')
		# Each box represents a part of the image where a particular object was detected.
		detection_boxes = detection_graph.get_tensor_by_name('detection_boxes:0')
		# Each score represent how level of confidence for each of the objects.
		# Score is shown on the result image, together with the class label.
		detection_scores = detection_graph.get_tensor_by_name('detection_scores:0')
		detection_classes = detection_graph.get_tensor_by_name('detection_classes:0')
		num_detections = detection_graph.get_tensor_by_name('num_detections:0')
		# the array based representation of the image data into the variable, this will be used later in order to prepare the
		# result image with boxes and labels on it.
		image_np = data
#		image_np = load_image_into_numpy_array(image)
		# Expand dimensions since the model expects images to have shape: [1, None, None, 3]
		image_np_expanded = np.expand_dims(image_np, axis=0)
		# Actual detection.
		(boxes, scores, classes, num) = sess.run(
		[detection_boxes, detection_scores, detection_classes, num_detections],
		feed_dict={image_tensor: image_np_expanded})
		# Visualization of the results of a detection.
		vis_util.visualize_boxes_and_labels_on_image_array(
		image_np,
		np.squeeze(boxes),
		np.squeeze(classes).astype(np.int32),
		np.squeeze(scores),
		category_index,
		use_normalized_coordinates=True,
		line_thickness=8)
		return image_np


src/imagenet_server_node.py
#! /usr/bin/env python

import rospy

import actionlib
import std_msgs.msg
from sensor_msgs.msg import CompressedImage
import imagenet_server.msg
import sys

class ImageNetServer():
	def __init__(self, sub_topic):
		self._sub_topic = sub_topic

		self._as = actionlib.SimpleActionServer("image_net_server",
                                            imagenet_server.msg.ImageNetAction,
                                            execute_cb=self.execute_cb,
                                            auto_start = False)

		self.image_sub = rospy.Subscriber(self._sub_topic, CompressedImage, self.receiveImage, queue_size = 1, buff_size=1000000000)
		rospy.loginfo("ImageNet action server running.")
		self.image = CompressedImage()
		self._as.start();

	def receiveImage(self ,im_data):
		self.image = im_data

	def execute_cb(self, goal):
		rospy.loginfo("Goal Received!")
		result = imagenet_server.msg.ImageNetResult()
		result.image = self.image
		try:
			self._as.set_succeeded(result)
		except Exception as e:
			rospy.logerr(str(e))
			self._as.set_aborted(text=str(e))

if __name__ == '__main__':
	rospy.init_node('ImageNet_action_server_node')
	sub_topic = "/xtion/rgb/image_rect_color/compressed"
#	sub_topic = "/usb_cam/image_raw/compressed"
	server = ImageNetServer(sub_topic)
	rospy.spin()


src/obj_det_fun_advanced.py
import numpy as np
import os
import six.moves.urllib as urllib
import sys
import tarfile
import tensorflow as tf
import zipfile
import rospy
import getpass 

from collections import defaultdict
from io import StringIO
from matplotlib import pyplot as plt
from PIL import Image

#username = getpass.getuser()
TF_MODELS_PATH = '/TFM/models/research/object_detection'
MODEL_DATABASE = '/TFM/model_database'
		
# This is needed to access files from research folder of Tensorflow models API
sys.path.append(TF_MODELS_PATH)	

from utils import label_map_util

from utils import visualization_utils as vis_util
	
# What model to download or use.
#MODEL_NAME = 'ssd_inception_v2_coco_11_06_2017'
MODEL_NAME = 'ssd_inception_v2_items'
MODEL_FILE = MODEL_NAME + '.tar.gz'
DOWNLOAD_BASE = 'http://download.tensorflow.org/models/object_detection/'

#local path directory of all models
MODEL_PATH = os.path.join(MODEL_DATABASE, MODEL_FILE)

# Path to frozen detection graph. This is the actual model that is used for the object detection.
PATH_TO_CKPT = MODEL_DATABASE + '/' + MODEL_NAME + '/frozen_inference_graph.pb'

# List of the strings that is used to add correct label for each box.
#PATH_TO_LABELS = os.path.join(TF_MODELS_PATH + '/data', 'mscoco_label_map.pbtxt')
PATH_TO_LABELS = os.path.join(MODEL_DATABASE, MODEL_NAME, 'items_label_map.pbtxt')

#NUM_CLASSES = 90
NUM_CLASSES = 3

def loadmodel():
	if os.path.exists(PATH_TO_CKPT):
	    rospy.logwarn('Found and verified .pb file : ' +  PATH_TO_CKPT)	
	    return
	if os.path.exists(MODEL_PATH):
	    rospy.logwarn('Found and verified : ' +  MODEL_PATH)
	    tar_file = tarfile.open(MODEL_PATH)
	else:
	    rospy.logwarn('The model is not available at the model database location: '+ MODEL_PATH)
	    rospy.logwarn('Attempting to download over internet:' + MODEL_FILE) 
	    opener = urllib.request.URLopener()
	    opener.retrieve(DOWNLOAD_BASE + MODEL_FILE, MODEL_FILE)
	    rospy.logwarn('\nDownload Complete!')
	    tar_file = tarfile.open(MODEL_FILE)
	for file in tar_file.getmembers():
	  file_name = os.path.basename(file.name)
	  if 'frozen_inference_graph.pb' in file_name:
	    #choose this option only if you want to extract in current work directory cwd
#		    tar_file.extract(file, os.getcwd())
	    tar_file.extract(file, MODEL_DATABASE)

def objdetect():
	detection_graph = tf.Graph()
	with detection_graph.as_default():
	  od_graph_def = tf.GraphDef()
	  with tf.gfile.GFile(PATH_TO_CKPT, 'rb') as fid:
	    serialized_graph = fid.read()
	    od_graph_def.ParseFromString(serialized_graph)
	    tf.import_graph_def(od_graph_def, name='')
	label_map = label_map_util.load_labelmap(PATH_TO_LABELS)
	categories = label_map_util.convert_label_map_to_categories(label_map, max_num_classes=NUM_CLASSES, use_display_name=True)
	category_index = label_map_util.create_category_index(categories)
	return detection_graph, category_index

def load_image_into_numpy_array(image):
#	(im_width, im_height) = image.size
	(im_width, im_height, _) = image.shape
#	return np.array(image.getdata()).reshape((im_height, im_width, 3)).astype(np.uint8)
	return np.array(image).reshape((im_height, im_width, 3)).astype(np.uint8)

def detect(sess, detection_graph,category_index, data):
	IMAGE_SIZE = (12, 8)
	with detection_graph.as_default():
		# Definite input and output Tensors for detection_graph
		image_tensor = detection_graph.get_tensor_by_name('image_tensor:0')
		# Each box represents a part of the image where a particular object was detected.
		detection_boxes = detection_graph.get_tensor_by_name('detection_boxes:0')
		# Each score represent how level of confidence for each of the objects.
		# Score is shown on the result image, together with the class label.
		detection_scores = detection_graph.get_tensor_by_name('detection_scores:0')
		detection_classes = detection_graph.get_tensor_by_name('detection_classes:0')
		num_detections = detection_graph.get_tensor_by_name('num_detections:0')
		# the array based representation of the image data into the variable, this will be used later in order to prepare the
		# result image with boxes and labels on it.
		image_np = data
#		image_np = load_image_into_numpy_array(image)
		# Expand dimensions since the model expects images to have shape: [1, None, None, 3]
		image_np_expanded = np.expand_dims(image_np, axis=0)
		# Actual detection.
		(boxes, scores, classes, num) = sess.run(
		[detection_boxes, detection_scores, detection_classes, num_detections],
		feed_dict={image_tensor: image_np_expanded})

		# Minimum score threshold
		minimum_threshold = 0.55 #default value is 0.5 (50%) used by the API

		# Visualization of the results of a detection.
		vis_util.visualize_boxes_and_labels_on_image_array(
		image_np,
		np.squeeze(boxes),
		np.squeeze(classes).astype(np.int32),
		np.squeeze(scores),
		category_index,
		use_normalized_coordinates=True,
		min_score_thresh=minimum_threshold,
		line_thickness=8)

		# Extraction of the detection scores, classes, boxes and number
		boxes = np.squeeze(boxes)
		scores = np.squeeze(scores)
		classes = np.squeeze(classes).astype(np.int32)

		detected_boxes = []
		detected_classes = []
		detected_scores = []
		num_detected = 0
		(im_width, im_height,_) = image_np.shape
		for i in range(int(boxes.shape[0])):
			if scores[i] > minimum_threshold:
				num_detected = num_detected + 1
				box = (boxes[i].tolist())
				box[0] = int(np.ceil(box[0]*im_width))
				box[2] = int(np.ceil(box[2]*im_width))
				box[1] = int(np.ceil(box[1]*im_height))
				box[3] = int(np.ceil(box[3]*im_height))
				detected_boxes.append(tuple(box))
				detected_classes.append(category_index.get(classes[i]).get('name').encode('utf8'))
				detected_scores.append(scores[i])

#		print('detected classes : {}, detected_scores : {}, num_detections : {}, detected boxes : {}'.format(detected_classes, detected_scores, num_detected, detected_boxes))
#		print('box type {} , shape {}'.format(type(detected_boxes), len(detected_boxes)))
		return image_np, num_detected, detected_classes, detected_scores, detected_boxes


src/obj_det_class.py
import numpy as np
import os
import six.moves.urllib as urllib
import sys
import tarfile
import tensorflow as tf
import zipfile
import rospy
import yaml

from object_detection.utils import label_map_util
from object_detection.utils import visualization_utils as vis_util

from collections import defaultdict
from io import StringIO
from matplotlib import pyplot as plt
from PIL import Image

DOWNLOAD_BASE = 'http://download.tensorflow.org/models/object_detection/'

class ObjectDetectionAPI():
	def __init__(self, model_name = "ssd_inception_v2_coco_11_06_2017", path_to_the_model_database = "/TFM/model_database", path_to_tf_models = "/TFM/models"):
		self.model_name = model_name
		self.path_to_the_model_database = path_to_the_model_database
		self.path_to_tf_models = path_to_tf_models

		# This function looks for the models in the location MODEL_DATABASE, if not it downloads them over internet
		# if they are standard models
		self.look_for_the_model(self.model_name, self.path_to_the_model_database)

		# Loads the detection graph and the category index from the extracted info from the look_for_the_model method
		self.load_detection_graph_and_category_index()

		# This config helps to run the tensorflow without having some memory issues
		# For instance, when cuda is not able to allocate some memory.
		self.config = tf.ConfigProto()
		self.config.gpu_options.allow_growth = True
		self.session = tf.Session(config = self.config, graph = self.detection_graph)
		rospy.loginfo('Tensorflow inference session successfully created!!');

	def __del__(self):
		self.session.close()
		rospy.loginfo('Tensorflow inference session successfully closed!!')

	def look_for_the_model(self, model_name, path_to_the_model_database):
		file_suffix = ".tar.gz"
		self.model_path = os.path.join(path_to_the_model_database, model_name)
		self.model_file = self.model_path + file_suffix
		#self.path_to_the_inference_graph = os.path.join(path_to_the_model_database, model_name, 'frozen_inference_graph.pb')
		#self.path_to_label_map_file = self.model_path + "/*_label_map.pbtxt"
		self.path_to_yaml = os.path.join(self.model_path, 'config.yaml')
		if os.path.exists(self.path_to_yaml):
			with open(self.path_to_yaml, "r") as file:
				config_data = yaml.load(file)
				self.path_to_label_map_file = os.path.join(path_to_the_model_database, model_name, config_data["label_map_file"])
				self.path_to_the_inference_graph = os.path.join(path_to_the_model_database, model_name, config_data["inference_graph"])

			if os.path.exists(self.path_to_the_inference_graph) and os.path.exists(self.path_to_label_map_file):
			    rospy.logwarn('Found and verified .pb file : ' +  self.path_to_the_inference_graph)	
			    rospy.logwarn('Found the label_map file : ' +  os.path.basename(self.path_to_label_map_file))
			    return
		else:
			self.path_to_label_map_file = os.path.join(self.path_to_tf_models, 'research/object_detection/data', 'mscoco_label_map.pbtxt')
			self.path_to_the_inference_graph = os.path.join(path_to_the_model_database, model_name, 'frozen_inference_graph.pb')
	    	if os.path.exists(self.path_to_the_inference_graph) and os.path.exists(self.path_to_label_map_file):
			    rospy.logwarn('Found and verified .pb file : ' +  self.path_to_the_inference_graph)	
			    rospy.logwarn('Found the label_map file : ' +  os.path.basename(self.path_to_label_map_file))
			    return

		if os.path.exists(self.model_file):
		    rospy.logwarn('Found and verified : ' +  self.model_file)
		    tar_file = tarfile.open(self.model_file)
		else:
		    rospy.logwarn('The model is not available at the model database location: '+ self.model_file)
		    rospy.logwarn('Attempting to download over internet:' + model_name + file_suffix) 
		    opener = urllib.request.URLopener()
		    opener.retrieve(DOWNLOAD_BASE + model_name + file_suffix, model_name + file_suffix)
		    rospy.logwarn('\nDownload Complete!')
		    tar_file = tarfile.open(self.model_file)

		# Getting the config yaml
		for file in tar_file.getmembers():
			file_name = os.path.basename(file.name)
			if 'config.yaml' in file_name:
				##choose this option only if you want to extract in current work directory cwd
				#tar_file.extract(file, os.getcwd())
				tar_file.extract(file, path_to_the_model_database)
				self.path_to_yaml = os.path.join(self.model_path, 'config.yaml')
				with open(self.path_to_yaml, "r") as file:
					config_data = yaml.load(file)
					self.path_to_label_map_file = os.path.join(path_to_the_model_database, model_name, config_data["label_map_file"])
					self.path_to_the_inference_graph = os.path.join(path_to_the_model_database, model_name, config_data["inference_graph"])
					break

		for file in tar_file.getmembers():
			file_name = os.path.basename(file.name)
			if os.path.basename(self.path_to_the_inference_graph) in file_name:
				tar_file.extract(file, path_to_the_model_database)
			if os.path.basename(self.path_to_label_map_file) in file_name:
				tar_file.extract(file, path_to_the_model_database)

	def load_detection_graph_and_category_index(self):
		self.detection_graph = tf.Graph()
		with self.detection_graph.as_default():
			od_graph_def = tf.GraphDef()
			with tf.gfile.GFile(self.path_to_the_inference_graph, 'rb') as fid:
				serialized_graph = fid.read()
				od_graph_def.ParseFromString(serialized_graph)
				tf.import_graph_def(od_graph_def, name='')
		label_map = label_map_util.load_labelmap(self.path_to_label_map_file)
		self.num_classes = label_map_util.get_max_label_map_index(label_map)
		categories = label_map_util.convert_label_map_to_categories(label_map, max_num_classes=self.num_classes, use_display_name=True)
		self.category_index = label_map_util.create_category_index(categories)
		#return detection_graph, category_index

	def get_detection_graph_and_category_index(self, model_name = "ssd_inception_v2_coco_11_06_2017", path_to_the_model_database = "/TFM/model_database"):
		self.look_for_the_model(model_name, path_to_the_model_database)
		self.detection_graph = tf.Graph()
		with self.detection_graph.as_default():
			od_graph_def = tf.GraphDef()
			with tf.gfile.GFile(self.path_to_the_inference_graph, 'rb') as fid:
				serialized_graph = fid.read()
				od_graph_def.ParseFromString(serialized_graph)
				tf.import_graph_def(od_graph_def, name='')
		label_map = label_map_util.load_labelmap(self.path_to_label_map_file)
		self.num_classes = label_map_util.get_max_label_map_index(label_map)
		categories = label_map_util.convert_label_map_to_categories(label_map, max_num_classes=self.num_classes, use_display_name=True)
		self.category_index = label_map_util.create_category_index(categories)
		return detection_graph, category_index

	def update_model(self, model_name = "ssd_inception_v2_coco_11_06_2017", path_to_the_model_database = "/TFM/model_database"):
		self.model_name = model_name
		self.path_to_the_model_database = path_to_the_model_database
		self.look_for_the_model(self.model_name, self.path_to_the_model_database)
		self.load_detection_graph_and_category_index()
		self.create_new_session()

	def load_image_into_numpy_array(self, image):
		(self.im_width, self.im_height, _) = image.shape
		return np.array(image).reshape((im_height, im_width, 3)).astype(np.uint8)
		#(im_width, im_height) = image.size
		#return np.array(image.getdata()).reshape((im_height, im_width, 3)).astype(np.uint8)

	def create_new_session(self, new_detection_graph, new_category_index):
		# the session need to be closed before, in order to release memory for other session
		self.session.close()
		self.detection_graph = new_detection_graph
		self.category_index = new_category_index
		self.session = tf.Session(config=self.config, graph = self.detection_graph)
		rospy.loginfo('Tensorflow new inference session successfully created!!');

	def replace_session(self, session, new_detection_graph, new_category_index):
		rospy.logwarn('You are replacing the sessions of the API, please be sure to close the session to release memory');
		self.detection_graph = new_detection_graph
		self.category_index = new_category_index
		self.session = session
		rospy.loginfo('Tensorflow inference session successfully replaced!!');

	def detect(self, data):
		with self.detection_graph.as_default():
			# Definite input and output Tensors for self.detection_graph
			image_tensor = self.detection_graph.get_tensor_by_name('image_tensor:0')
			# Each box represents a part of the image where a particular object was detected.
			detection_boxes = self.detection_graph.get_tensor_by_name('detection_boxes:0')
			# Each score represent how level of confidence for each of the objects.
			# Score is shown on the result image, together with the class label.
			detection_scores = self.detection_graph.get_tensor_by_name('detection_scores:0')
			detection_classes = self.detection_graph.get_tensor_by_name('detection_classes:0')
			num_detections = self.detection_graph.get_tensor_by_name('num_detections:0')
			# the array based representation of the image data into the variable, this will be used later in order to prepare the
			# result image with boxes and labels on it.
			image_np = data
	#		image_np = load_image_into_numpy_array(image)
			# Expand dimensions since the model expects images to have shape: [1, None, None, 3]
			image_np_expanded = np.expand_dims(image_np, axis=0)
			# Actual detection.
			(boxes, scores, classes, num) = self.session.run(
			[detection_boxes, detection_scores, detection_classes, num_detections],
			feed_dict={image_tensor: image_np_expanded})

			# Minimum score threshold
			minimum_threshold = 0.55 #default value is 0.5 (50%) used by the API

			# Visualization of the results of a detection.
			vis_util.visualize_boxes_and_labels_on_image_array(
			image_np,
			np.squeeze(boxes),
			np.squeeze(classes).astype(np.int32),
			np.squeeze(scores),
			self.category_index,
			use_normalized_coordinates=True,
			min_score_thresh=minimum_threshold,
			line_thickness=8)

			# Extraction of the detection scores, classes, boxes and number
			boxes = np.squeeze(boxes)
			scores = np.squeeze(scores)
			classes = np.squeeze(classes).astype(np.int32)

			detected_boxes = []
			detected_classes = []
			detected_scores = []
			num_detected = 0
			(self.im_width, self.im_height,_) = image_np.shape
			for i in range(int(boxes.shape[0])):
				if scores[i] > minimum_threshold:
					num_detected = num_detected + 1
					box = (boxes[i].tolist())
					box[0] = int(np.ceil(box[0]*self.im_width))
					box[2] = int(np.ceil(box[2]*self.im_width))
					box[1] = int(np.ceil(box[1]*self.im_height))
					box[3] = int(np.ceil(box[3]*self.im_height))
					detected_boxes.append(tuple(box))
					detected_classes.append(self.category_index.get(classes[i]).get('name').encode('utf8'))
					detected_scores.append(scores[i])

			return image_np, num_detected, detected_classes, detected_scores, detected_boxes
'''
	def detect(self, sess, detection_graph, category_index, data):
		IMAGE_SIZE = (12, 8)
		with detection_graph.as_default():
			# Definite input and output Tensors for detection_graph
			image_tensor = detection_graph.get_tensor_by_name('image_tensor:0')
			# Each box represents a part of the image where a particular object was detected.
			detection_boxes = detection_graph.get_tensor_by_name('detection_boxes:0')
			# Each score represent how level of confidence for each of the objects.
			# Score is shown on the result image, together with the class label.
			detection_scores = detection_graph.get_tensor_by_name('detection_scores:0')
			detection_classes = detection_graph.get_tensor_by_name('detection_classes:0')
			num_detections = detection_graph.get_tensor_by_name('num_detections:0')
			# the array based representation of the image data into the variable, this will be used later in order to prepare the
			# result image with boxes and labels on it.
			image_np = data
	#		image_np = load_image_into_numpy_array(image)
			# Expand dimensions since the model expects images to have shape: [1, None, None, 3]
			image_np_expanded = np.expand_dims(image_np, axis=0)
			# Actual detection.
			(boxes, scores, classes, num) = sess.run(
			[detection_boxes, detection_scores, detection_classes, num_detections],
			feed_dict={image_tensor: image_np_expanded})

			# Minimum score threshold
			minimum_threshold = 0.55 #default value is 0.5 (50%) used by the API

			# Visualization of the results of a detection.
			vis_util.visualize_boxes_and_labels_on_image_array(
			image_np,
			np.squeeze(boxes),
			np.squeeze(classes).astype(np.int32),
			np.squeeze(scores),
			category_index,
			use_normalized_coordinates=True,
			min_score_thresh=minimum_threshold,
			line_thickness=8)

			# Extraction of the detection scores, classes, boxes and number
			boxes = np.squeeze(boxes)
			scores = np.squeeze(scores)
			classes = np.squeeze(classes).astype(np.int32)

			detected_boxes = []
			detected_classes = []
			detected_scores = []
			num_detected = 0
			(self.im_width, self.im_height,_) = image_np.shape
			for i in range(int(boxes.shape[0])):
				if scores[i] > minimum_threshold:
					num_detected = num_detected + 1
					box = (boxes[i].tolist())
					box[0] = int(np.ceil(box[0]*self.im_width))
					box[2] = int(np.ceil(box[2]*self.im_width))
					box[1] = int(np.ceil(box[1]*self.im_height))
					box[3] = int(np.ceil(box[3]*self.im_height))
					detected_boxes.append(tuple(box))
					detected_classes.append(category_index.get(classes[i]).get('name').encode('utf8'))
					detected_scores.append(scores[i])

	#		print('detected classes : {}, detected_scores : {}, num_detections : {}, detected boxes : {}'.format(detected_classes, detected_scores, num_detected, detected_boxes))
	#		print('box type {} , shape {}'.format(type(detected_boxes), len(detected_boxes)))
			return image_np, num_detected, detected_classes, detected_scores, detected_boxes
'''

'''
	def look_for_the_model(self):
		file_suffix = ".tar.gz"
		self.model_path = os.path.join(self.path_to_the_model_database, self.model_name)
		self.model_file = self.model_path + file_suffix
		#self.path_to_the_inference_graph = os.path.join(self.path_to_the_model_database, self.model_name, 'frozen_inference_graph.pb')
		#self.path_to_label_map_file = self.model_path + "/*_label_map.pbtxt"
		self.path_to_yaml = os.path.join(self.model_path, 'config.yaml')
		if os.path.exists(self.path_to_yaml):
			with open(self.path_to_yaml, "r") as file:
				config_data = yaml.load(file)
				self.path_to_label_map_file = os.path.join(self.path_to_the_model_database, self.model_name, config_data["label_map_file"])
				self.path_to_the_inference_graph = os.path.join(self.path_to_the_model_database, self.model_name, config_data["inference_graph"])

			if os.path.exists(self.path_to_the_inference_graph) and os.path.exists(self.path_to_label_map_file):
			    rospy.logwarn('Found and verified .pb file : ' +  self.path_to_the_inference_graph)	
			    rospy.logwarn('Found the label_map file : ' +  os.path.basename(self.path_to_label_map_file))
			    return
		if os.path.exists(self.model_file):
		    rospy.logwarn('Found and verified : ' +  self.model_file)
		    tar_file = tarfile.open(self.model_file)
		else:
		    rospy.logwarn('The model is not available at the model database location: '+ self.model_file)
		    rospy.logwarn('Attempting to download over internet:' + self.model_name + file_suffix) 
		    opener = urllib.request.URLopener()
		    opener.retrieve(DOWNLOAD_BASE + self.model_name + file_suffix, self.model_name + file_suffix)
		    rospy.logwarn('\nDownload Complete!')
		    tar_file = tarfile.open(self.model_file)

		# Getting the config yaml
		for file in tar_file.getmembers():
			file_name = os.path.basename(file.name)
			if 'config.yaml' in file_name:
				##choose this option only if you want to extract in current work directory cwd
				#tar_file.extract(file, os.getcwd())
				tar_file.extract(file, self.path_to_the_model_database)
				self.path_to_yaml = os.path.join(self.model_path, 'config.yaml')
				with open(self.path_to_yaml, "r") as file:
					config_data = yaml.load(file)
					self.path_to_label_map_file = os.path.join(self.path_to_the_model_database, self.model_name, config_data["label_map_file"])
					self.path_to_the_inference_graph = os.path.join(self.path_to_the_model_database, self.model_name, config_data["inference_graph"])

		for file in tar_file.getmembers():
			file_name = os.path.basename(file.name)
			if os.path.basename(self.path_to_the_inference_graph) in file_name:
				tar_file.extract(file, self.path_to_the_model_database)
			if os.path.basename(self.path_to_label_map_file) in file_name:
				tar_file.extract(file, self.path_to_the_model_database)
'''